daq/runner.py

```diff
@@ -790,10 +791,6 @@ def _target_set_finalize(self, device, result_set, reason):
                 self._linger_exit = 1
         self._result_sets[device] = result_set
 
-        if self._device_result_client:
```

Comment:
This can be restored to replace line 771 and 814-816 when directly storing test_results on the device class; and self._send_device_results can be called with just 1 argument “device”.

Reply:
I think the problem is that in my case the test_results is processed in `_target_set_cancel()`, which is called after `_target_set_finalize()`, which means when calling `_target_set_finalize()`, the test_result is not ready yet.

Reply:
You can modify where the the device.test_results are populated; for example, when there are no more tests remaining under that host. That way you have access to device.results under _target_set_finalize. Also the current logic where it returns PortBehavior.failed when any test module fails should also be left intact. 
Devices in host and in runner are the same objects; devices are passed around everywhere to reference information related to themselves.

Reply:
I checked the logic again and I don't think it is possible to move the report processing to _run_next_test() without chaning the way how the test is terminated in case of exception. I added some new logic on top of my changes to handle the exception case. PTAL..